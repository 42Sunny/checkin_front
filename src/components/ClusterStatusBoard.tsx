import React, { useEffect, useState } from "react";
import Circle from "./Circle";
import classes from "../styles/components/ClusterStatusBoard.module.css";
import useCluster from "../utils/hooks/useCluster";

const ClusterStatusBoard = () => {
  const {
    cluster: { gaepo, gaepoLimitation, seocho, seochoLimitation },
  } = useCluster();

  const [clusterCongestion, setClusterCongestion] = useState({
    // 개포, 서초 인원 수에 따른 혼잡도
    gaepoCongestion: "",
    seochoCongestion: "",
  });
  const { gaepoCongestion, seochoCongestion } = clusterCongestion;

  const getClusterCongestion = (current: number, max: number) => {
    const state = current / max;

    if (state > 0.8) {
      return "red";
    }
    if (state <= 0.8 && state > 0.4) {
      return "orange";
    }
    return "green";
  };

  useEffect(() => {
    setClusterCongestion({
      gaepoCongestion: getClusterCongestion(gaepo, gaepoLimitation),
      seochoCongestion: getClusterCongestion(seocho, seochoLimitation),
    });
  }, [gaepo, gaepoLimitation, seocho, seochoLimitation]);

  return (
    <div className={classes.wrap}>
      <div className={classes.item}>
        <span className={classes.title}>
          개포
          <span className={classes.count}>
            <span>
              {gaepo} / {gaepoLimitation}
            </span>
          </span>
          <Circle color={gaepoCongestion} />
        </span>
      </div>
      <div className={classes.item}>
        <span className={classes.title}>
          서초
          <span className={classes.count}>
            <span>
              {seocho} / {seochoLimitation}
            </span>
          </span>
          <Circle color={seochoCongestion} />
        </span>
      </div>
    </div>
  );
};

export default ClusterStatusBoard;
