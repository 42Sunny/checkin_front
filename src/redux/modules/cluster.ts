import { createAction, ActionType } from "typesafe-actions";
import { formatOfficeHours } from "../../utils/time";

// actions
const SET_CLUSTER = "cluster/SET_CLUSTER";
const SET_CURRENT_USER_COUNT = "cluster/SET_CURRENT_USER_COUNT";
const SET_OFFICE_HOUR = "cluster/SET_OFFICE_HOUR";
const SET_LUNCH_TIME = "cluster/SET_LUNCH_TIME";
// action creators

export const setCluster =
  createAction(SET_CLUSTER)<Omit<Cluster, "officeHour" | "officeLunchTime">>();
export const setCurrentUserCount =
  createAction(SET_CURRENT_USER_COUNT)<{ gaepo: number; seocho: number }>();
export const setOfficeHour = createAction(SET_OFFICE_HOUR)<{ officeHour: string }>();
export const setLunchTime = createAction(SET_LUNCH_TIME)<{ officeLunchTime: string }>();
// type
const actions = { setCluster, setCurrentUserCount, setOfficeHour, setLunchTime };
type ClusterActions = ActionType<typeof actions>;

// initialState
const initialState: Cluster = {
  openAt: "",
  closeAt: "",
  seocho: 0,
  gaepo: 0,
  seochoLimitation: 0,
  gaepoLimitation: 0,
  officeHour: "",
  officeLunchTime: "",
};

// reducer
const cluster = (state = initialState, action: ClusterActions): Cluster => {
  switch (action.type) {
    case SET_CLUSTER: {
      const { openAt, closeAt, gaepo, seocho, gaepoLimitation, seochoLimitation } = action.payload;
      return {
        ...state,
        openAt,
        closeAt,
        gaepoLimitation,
        seochoLimitation,
        gaepo,
        seocho,
      };
    }
    case SET_CURRENT_USER_COUNT: {
      const { gaepo, seocho } = action.payload;
      return { ...state, gaepo, seocho };
    }
    case SET_OFFICE_HOUR: {
      const { officeHour } = action.payload;
      return { ...state, officeHour };
    }
    case SET_LUNCH_TIME: {
      const { officeLunchTime } = action.payload;
      return { ...state, officeLunchTime };
    }
    default:
      return state;
  }
};

export default cluster;
