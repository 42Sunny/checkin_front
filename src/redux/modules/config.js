// actions
const SET_CONFIG = 'SET_CONFIG';

// action creators
export const setConfig = data => {
  const { openAt, closeAt, seocho, gaepo } = data;
  return {
    type: SET_CONFIG,
    payload: {
      openAt: openAt,
      closeAt: closeAt,
      seocho: seocho,
      gaepo: gaepo
    }
  };
};

// initalState
const initalState = {
  openAt: null,
  closeAt: null,
  seocho: 0,
  gaepo: 0
};

// reducer
export const config = (state = initalState, action) => {
  switch (action.type) {
    case SET_CONFIG:
      return {
        ...state,
        openAt: action.payload.openAt,
        closeAt: action.payload.closeAt,
        seocho: action.payload.seocho,
        gaepo: action.payload.gaepo
      };
    default:
      return state;
  }
};
