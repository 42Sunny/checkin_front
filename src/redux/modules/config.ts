import { createAction, ActionType } from "typesafe-actions";

// actions
const SET_CONFIG = "config/SET_CONFIG";
const SET_CURRENT_USER_COUNT = "config/SET_CURRENT_USER_COUNT";
// action creators

export const setConfig = createAction(SET_CONFIG)<Config>();
export const setCurrentUserCount = createAction(SET_CURRENT_USER_COUNT)<{ gaepo: number; seocho: number }>();

// type
const actions = { setConfig, setCurrentUserCount };
type ConfigActions = ActionType<typeof actions>;

// initalState
const initalState: Config = {
  openAt: "",
  closeAt: "",
  seocho: 0,
  gaepo: 0,
  seochoLimitation: 0,
  gaepoLimitation: 0,
};

// reducer
const configReducer = (state = initalState, action: ConfigActions): Config => {
  switch (action.type) {
    case SET_CONFIG: {
      const { openAt, closeAt, gaepo, seocho, gaepoLimitation, seochoLimitation } = action.payload;
      return { ...state, openAt, closeAt, gaepoLimitation, seochoLimitation, gaepo, seocho };
    }
    case SET_CURRENT_USER_COUNT: {
      const { gaepo, seocho } = action.payload;
      return { ...state, gaepo, seocho };
    }
    default:
      return state;
  }
};

export default configReducer;
