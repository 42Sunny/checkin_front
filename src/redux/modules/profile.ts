import * as Sentry from "@sentry/react";
import { ActionType, createAction, Reducer } from "typesafe-actions";
import DEFAULT_PROFILE from "../../assets/user-default.png";
// actions
const LOGIN = "profile/LOGIN";
const LOGOUT = "profile/LOGOUT";
const SET_USER = "profile/SET_USER";
const SET_AUTH = "profile/SET_AUTH";

// action creators
export const login = createAction(LOGIN)();
export const logout = createAction(LOGOUT)();
export const setUser = createAction(SET_USER)<Omit<User, "isLogin" | "isAdmin">>();
export const setAuth = createAction(SET_AUTH)<{ isAdmin: boolean }>();

// type
const actions = { setUser, login, logout, setAuth };
type UserActions = ActionType<typeof actions>;

// initialState
export const initialState: User = {
  isLogin: false,
  id: "",
  state: "checkOut",
  checkinAt: null,
  checkoutAt: null,
  profile: DEFAULT_PROFILE,
  isAdmin: false,
};

// reducer
const user: Reducer<User, UserActions> = (state = initialState, action): User => {
  switch (action.type) {
    case LOGIN:
      return {
        ...state,
        isLogin: true,
      };
    case LOGOUT:
      return {
        ...state,
        isLogin: false,
      };
    case SET_AUTH: {
      const { isAdmin } = action.payload;
      return { ...state, isAdmin };
    }
    case SET_USER: {
      const { checkinAt, checkoutAt, id, profile, state: userState } = action.payload;
      Sentry.setUser({ username: id });
      return { ...state, id, state: userState, checkinAt, checkoutAt, profile };
    }

    default:
      return state;
  }
};

export default user;
