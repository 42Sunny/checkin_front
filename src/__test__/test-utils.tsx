// test-utils.jsx
import { configureStore } from "@reduxjs/toolkit";
import { Queries, render, RenderResult } from "@testing-library/react";
import { createMemoryHistory } from "history";
import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
// Import your own reducer
import renderer from "react-test-renderer";
import appStore from "../redux/configureStore";
import rootReducer, { RootState } from "../redux/modules";
import { initialState as clusterInitState } from "../redux/modules/cluster";
import { initialState as userInitState } from "../redux/modules/user";

const makeCustomStore = ({
  cluster,
  user,
}: {
  cluster?: typeof clusterInitState;
  user?: typeof userInitState;
} = {}) => {
  return {
    clusterReducer: { ...clusterInitState, ...cluster },
    userReducer: { ...userInitState, ...user },
  };
};

const customRender = (
  ui: ReactElement,
  preloadedState: RootState = makeCustomStore(),
  { store = configureStore({ reducer: rootReducer, preloadedState }), ...renderOptions } = {},
  history = createMemoryHistory(),
): RenderResult<Queries, HTMLElement> => {
  const Wrapper: any = ({ children }: { children: ReactElement }) => {
    return (
      <Provider store={store}>
        <Router history={history}>{children}</Router>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

const createWithWrapper = (children: ReactElement) =>
  renderer.create(
    <Provider store={appStore}>
      <Router history={createMemoryHistory()}>{children}</Router>
    </Provider>,
  );

// re-export everything
export * from "@testing-library/react";
// override render method
export { customRender as render, createWithWrapper, makeCustomStore };
