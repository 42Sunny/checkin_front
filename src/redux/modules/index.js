import { combineReducers } from 'redux';
import { user } from './user';
import { config } from './config';
import { status } from './status';

const rootReducer = combineReducers({
  user: user,
  config: config,
  status: status
});

export default rootReducer;
