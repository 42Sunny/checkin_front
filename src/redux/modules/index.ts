import { combineReducers } from "redux";
import userReducer from "./user";
import configReducer from "./config";
import statusReducer from "./status";

const rootReducer = combineReducers({
  userReducer,
  configReducer,
  statusReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
