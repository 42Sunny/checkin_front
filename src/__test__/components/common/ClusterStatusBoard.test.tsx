import ClusterStatusBoard from "../../../components/common/ClusterStatusBoard";
import { initialState as clusterInitState } from "../../../redux/modules/cluster";
import { createWithWrapper, makeCustomStore, render, screen } from "../../test-utils";

describe("src/components/common/ClusterStatusBoard.tsx", () => {
  it("matched snapshot", () => {
    const component = createWithWrapper(<ClusterStatusBoard />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("renders cadet's count", () => {
    const cadetCounts = {
      gaepo: 10,
      gaepoLimitation: 10,
      seocho: 4,
      seochoLimitation: 42,
    };
    render(
      <ClusterStatusBoard />,
      makeCustomStore({ cluster: { ...clusterInitState, ...cadetCounts } }),
    );
    const gaepoCount = screen.getByTestId(/개포-cadet-count/i);
    const seochoCount = screen.getByTestId(/서초-cadet-count/i);
    expect(gaepoCount).toHaveTextContent(`${cadetCounts.gaepo} / ${cadetCounts.gaepoLimitation}`);
    expect(seochoCount).toHaveTextContent(
      `${cadetCounts.seocho} / ${cadetCounts.seochoLimitation}`,
    );
  });

  it("renders correct congestion", () => {
    const cadetCounts = {
      gaepo: 10,
      gaepoLimitation: 10,
      seocho: 4,
      seochoLimitation: 42,
    };

    render(
      <ClusterStatusBoard />,
      makeCustomStore({ cluster: { ...clusterInitState, ...cadetCounts } }),
    );
    const [gaepoCongestionEl, seochoCongestionEl] = screen.getAllByRole("status");
    const [gaepoCongestion, seochoCongestion] = [
      cadetCounts.gaepo / cadetCounts.gaepoLimitation,
      cadetCounts.seocho / cadetCounts.seochoLimitation,
    ];

    const correctCongestion = (congestion: any) => {
      if (congestion > 0.8) return "red";
      if (congestion > 0.4) return "orange";
      return "green";
    };
    expect(gaepoCongestionEl).toHaveClass(correctCongestion(gaepoCongestion));
    expect(seochoCongestionEl).toHaveClass(correctCongestion(seochoCongestion));
  });
});
