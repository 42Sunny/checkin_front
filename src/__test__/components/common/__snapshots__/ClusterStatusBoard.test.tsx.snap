// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/common/ClusterStatusBoard.tsx matched snapshot 1`] = `
<div
  className="wrap"
>
  <div
    className="item"
  >
    <span
      className="title"
    >
      개포
      <span
        className="count"
      >
        <span
          data-testid="개포-cadet-count"
        >
          0
           / 
          0
        </span>
      </span>
      <div
        className="circle green"
        role="status"
      />
    </span>
  </div>
  <div
    className="item"
  >
    <span
      className="title"
    >
      서초
      <span
        className="count"
      >
        <span
          data-testid="서초-cadet-count"
        >
          0
           / 
          0
        </span>
      </span>
      <div
        className="circle green"
        role="status"
      />
    </span>
  </div>
</div>
`;
